frontend/src/metabase/new_query/containers/TableSearch.jsx

```diff
@@ -0,0 +1,155 @@
+    props: {
+        getUrlForQuery: (StructuredQuery) => void,
+        backButtonUrl: string,
+
+        query: StructuredQuery,
+        metadata: Metadata,
+        metadataFetched: any,
+        fetchSegments: () => void,
+        fetchDatabases: () => void,
+        resetQuery: () => void
+    }
+
+    componentDidMount() {
+        this.props.fetchDatabases() // load databases if not loaded yet
+        this.props.fetchSegments(true) // segments may change more often so always reload them
+        this.props.resetQuery();
+    }
+
+    getUrlForTable = (table: Table) => {
+        const updatedQuery = this.props.query
+            .setDatabase(table.db)
+            .setTable(table)
+
+        return this.props.getUrlForQuery(updatedQuery);
+    }
+
+    getUrlForSegment = (segment: Segment) => {
+        const updatedQuery = this.props.query
+            .setDatabase(segment.table.database)
+            .setTable(segment.table)
+            .addFilter(segment.filterClause())
+
+        return this.props.getUrlForQuery(updatedQuery);
+    }
+
+    // TODO Atte Keinänen 9/11/17: How to differentiate tables and segments from each other?
+    getUrlForEntity = (segmentOrTable) => {
+        if (segmentOrTable.isTable) {
+            // Table object format doesn't match with other entity types
+            // so that's why the Table metadata object is stored in underlyingTable prop
+            return this.getUrlForTable(segmentOrTable.table)
+        } else {
+            // For segments we can use the object as-is
+            return this.getUrlForSegment(segmentOrTable)
+        }
+    }
+
+    getTableInEntitySearchFormat = (table, segmentsList) => {
+        const dbHasMultipleSchemas = table.db.schemaNames().length > 1
+        const getEntitySearchDescription = ({currentGrouping}) => {
+            const showDbName = currentGrouping.id !== "database"
+            const showSchemaName = dbHasMultipleSchemas && table.schema
+            return (
+                <span>
+                <b>{showDbName && table.db.name}</b>
+                {showDbName && showSchemaName && ' - '}
+                {showSchemaName && table.schema}
+            </span>
+            )
+        }
+
+        return {
+            name: table.display_name,
+            isTable: true,
+            table,
+            children:
+                // TODO Atte Keinänen 9/29/17: Should this be optimized, i.e. can the count of segments be high?
+                segmentsList
+                    .filter(segment => segment.table_id === table.id)
+                    .map(segment => Object.assign(segment, { getEntitySearchDescription })),
+            getEntitySearchDescription
+        }
+    }
+
+    render() {
+        const { backButtonUrl, metadataFetched, metadata } = this.props;
+
+        const isLoading = !metadataFetched.segments || !metadataFetched.databases
+
+        // TODO Atte Keinänen 8/22/17: If you call `/api/table/:id/table_metadata` it returns
```

Comment:
Things like this should bring in @senior or @camsaul  rather than continuing to paper over things.

Reply:
Just pinged @camsaul about this in Slack